File: app/src/main/java/tomaszkopacz/meetbam/tabs_service/PostAdapter.java
package tomaszkopacz.meetbam.tabs_service;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.Glide;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import tomaszkopacz.meetbam.R;
import tomaszkopacz.meetbam.activities.MainActivity;
import tomaszkopacz.meetbam.model.Post;

/**
 * Created by tomas on 03.03.2018.
 * Creates dynamic recycler view list of posts.
 */

public class PostAdapter extends RecyclerView.Adapter {

    // activity context
    private MainActivity activity;

    // data source
    private List<Post> posts = new ArrayList<>();

    // list
    private RecyclerView recyclerView;

    // photo base url
    private static final String BASE_URL = "http://meetbam.cba.pl/";

    /**
     * Constructor.
     */
    public PostAdapter(MainActivity activity, List<Post> posts, RecyclerView recyclerView){
        this.activity = activity;
        this.posts = posts;
        this.recyclerView = recyclerView;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.post_item, parent, false);

        // remove item when clicked
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int position = recyclerView.getChildAdapterPosition(view);
                posts.remove(position);
                notifyItemRemoved(position);
            }
        });

        return new PostViewHolder(view);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        Post post = posts.get(position);

        //set text views
        ((PostViewHolder) holder).name.setText(post.getName());
        ((PostViewHolder) holder).surname.setText(post.getSurname());
        ((PostViewHolder) holder).time.setText(post.getTime());

        // download and set image
        Glide
                .with(activity)
                .load(BASE_URL + post.getPhotoDir())
                .into(((PostViewHolder) holder).photo);

        // count time since post was uploaded:
        // original timestamp from server is "yyyy-MM-dd hh:mm:ss.xxxxxx", need to be
        // converted to "yyyy-MM-dd hh:mm:ss" pattern, so it can be used in
        // countTimeAgo(String timestamp) method
        String timestamp = post.getTime().substring(0, 19);
        long timeAgo = countTimeAgo(timestamp);

        // get appropriate info about how long ago post was uploaded
        String timeAgoText = getTimeAgoText(timeAgo);
        ((PostViewHolder) holder).time.setText(timeAgoText);

    }

    @Override
    public int getItemCount() {
        return posts.size();
    }

    /**
     * Counts time interval between now and given timestamp. The timestamp
     * should be given in "yyyy-MM-dd hh:mm:ss" pattern. The result is time in minutes.
     * @param timestamp
     * @return time interval started from timestamp and ending now. If error occurs, returns -1.
     */
    private long countTimeAgo(String timestamp){

        // date pattern
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss");

        // time interval
        long diff;

        try {
            // convert timestamp to given pattern
            Date timestampDate = simpleDateFormat.parse(timestamp);

            // get actual date and convert to pattern
            Date date = new Date();
            simpleDateFormat.format(date);

            // count difference in millis
            diff = date.getTime() - timestampDate.getTime();

            // convert millis to minutes
            diff = diff/(1000*60);

        } catch (ParseException e) {
            return -1;
        }

        return diff;
    }

    /**
     * Returns text according to a given time. If time is less than 1h, returns
     * result in minutes. If time is less than a day, returns result in hours. Otherwise,
     * returns result in days. If given time is less than zero, returns "TIME ERROR" message.
     * @param timeAgo time given in minutes
     * @return text about haw many time went off
     */
    private String getTimeAgoText(long timeAgo){

        if (timeAgo < 0)
            return "TIME ERROR";
        else if (timeAgo == 1)
            return "1" + " minute ago";
        else if (timeAgo < 60)
            return String.valueOf(timeAgo) + " minutes ago";
        else if (timeAgo < 120)
            return "1" + " hour ago";
        else if (Math.round(timeAgo) < 1440)
            return String.valueOf(timeAgo/60) + " hours ago";
        else if (timeAgo < 2880)
            return "1" + " day ago";
        else
            return String.valueOf(Math.round(timeAgo/(60*24))) + " days ago";
    }
}

File: app/src/main/java/tomaszkopacz/meetbam/view/PostAdapter.kt
package tomaszkopacz.meetbam.view

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.post_item.view.*
import tomaszkopacz.meetbam.R
import tomaszkopacz.meetbam.entity.Post
import tomaszkopacz.meetbam.interactor.StorageService
import tomaszkopacz.meetbam.presenter.RecyclerViewPresenter
import tomaszkopacz.meetbam.service.DateTimeCounter
import tomaszkopacz.meetbam.service.GlideApp


class PostAdapter(private val presenter: RecyclerViewPresenter)
    : RecyclerView.Adapter<PostAdapter.PostViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int)
            : PostViewHolder {

        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.post_item, parent, false)

        view.setOnClickListener{
            presenter.onItemClick(view)
        }

        view.setOnLongClickListener{
            presenter.onLongItemClick(view)
            true
        }

        return PostViewHolder(view)
    }

    override fun onBindViewHolder(holder: PostViewHolder, position: Int) {
        presenter.onItemBoundAtPosition(holder, position)
    }

    fun setContent(context: Context, holder: PostViewHolder, post: Post, service: StorageService) {
        holder.itemView.name1_textview.text = post.name1
        holder.itemView.name2_textview.text = post.name2
        holder.itemView.post_time_textview.text = post.time

        GlideApp
                .with(context)
                .load(service.getReference(post.url!!))
                .into(holder.itemView.post_imageview)

        val timeAgo = DateTimeCounter.countTimeAgo(post.time!!,
                DateTimeCounter.FORMAT_MINUTES)
        val timeAgoText = DateTimeCounter.getTimeAgoText(timeAgo,
                DateTimeCounter.FORMAT_MINUTES)
        holder.itemView.post_time_textview.text = context.getString(R.string.post_time, timeAgoText)
    }

    override fun getItemCount(): Int {
        return presenter.itemCount()
    }

    inner class PostViewHolder (itemView: View)
        : RecyclerView.ViewHolder(itemView)
}
